### PR TITLE
Pagination issue when pagination on server 

### PR DESCRIPTION
**Pull request by sarathdr**
[Issue](https://github.com/jbetancur/react-data-table-component/issues/331)

### CODE REVIEW

**Comment by jbetancur** [comment]
Thanks for taking a stab at fixing this.

I don't think you even need to re-calculate. Just detect if the paginationTotalRows changed then check the currentPage is > 1 and there are 0 records in the data.length or calculatedRows.length. then just move back a page

  if (pagination && paginationServer) {
    useDidUpdateEffect(() => {
      if (calculatedRows.length === 0 && currentPage > 1) {
        handleChangePage(currentPage - 1);
      }
    }, [paginationTotalRows]);
  }

I would also use `useDidUpdateEffect` since it doesn't fire on initial render

**Comment by jbetancur** [comment]
```
  if (pagination && paginationServer && ) {
    useDidUpdateEffect(() => {
      if (calculatedRows.length === 0 && currentPage > 1) {
        handleChangePage(currentPage - 1);
      }
    }, [paginationTotalRows]);
  }
```

**Comment by sarathdr** [comment]
> ```
>   if (pagination && paginationServer && ) {
>     useDidUpdateEffect(() => {
>       if (calculatedRows.length === 0 && currentPage > 1) {
>         handleChangePage(currentPage - 1);
>       }
>     }, [paginationTotalRows]);
>   }
> ```

`calculatedRows.length` this will be `> 0` since server should fetch the data for the last page? 
only the total changes?

File: src/DataTable/__tests__/DataTable.test.js [hunk 1]
```diff
@@ -1407,6 +1407,45 @@ describe('DataTable::Pagination', () => {
...
+        onChangePage={onChangePageMock}
+        paginationPerPage={1}
+        paginationTotalRows={1}
+        paginationRowsPerPageOptions={[1, 2]}
+        onChangeRowsPerPage={onChangeRowsPerPageMock}
+        paginationServer
+        pagination
+      />,
+    );
+
```
Comment (sarathdr) [inline src/DataTable/__tests__/DataTable.test.js]: @jbetancur Could you check the above props for re-render?

File: .gitignore [hunk 2]
```diff
@@ -68,3 +68,4 @@ storybook-static
 package-lock.json
```
Comment (sarathdr) [inline .gitignore]: should we also commit `package-lock.json` ? It seems snapshot is failing because of that

Comment (jbetancur) [reply .gitignore]: I would stick with yarn and not include this file.

**Comment by jbetancur** [comment]
caclulatedRows.length when using `paginationServer` should always be the same as data.length. It would probably be less confusing to just use data.length.

Since you are 'server-side" paging the calculatedRows.length will only be whatever the total length of the data loaded from you api is. So, if your paging for 10 records per page the length is 10.

What needs to be done is when the data.length reaches 0 then we trigger the actualy pagination back one page. e.g. you delete the last record on a page.

This is why `paginationTotalRows` matters when using paginationServer and why it should be set to whatever your pagination api returns as the total rows so that RDT can calculate what the total rows actually are instead of from data.length.

All of this works differently in client-side pagination since we have the entire dataset loaded which i why the code uses calculatedRows.length for those calculations (it uses Array.slice to give you the effect of pagination)

**Comment by jbetancur** [comment]
```
 if (pagination && paginationServer) {
    useDidUpdateEffect(() => {
      if (data.length === 0 && currentPage > 1) {
        handleChangePage(currentPage - 1);
      }
    }, [paginationTotalRows]);
  }
```

File: .gitignore (hunk 2 above)
Comment (sarathdr) [reply .gitignore]: Ah sorry I thought it uses npm

**Comment by sarathdr** [comment]
> ```
>  if (pagination && paginationServer) {
>     useDidUpdateEffect(() => {
>       if (data.length === 0 && currentPage > 1) {
>         handleChangePage(currentPage - 1);
>       }
>     }, [paginationTotalRows]);
>   }
> ```

ok I will test again and resubmit

File: src/DataTable/DataTable.js [hunk 3]
```diff
@@ -197,6 +197,18 @@ const DataTable = memo(({
     handleChangePage(recalculatedPage);
   }
 
+  if (paginationServer) {
```
Comment (janhartmann) [inline src/DataTable/DataTable.js]: Please see https://reactjs.org/docs/hooks-overview.html#rules-of-hooks

> Only call Hooks at the top level. Don’t call Hooks inside loops, conditions, or nested functions.

Comment (jbetancur) [reply src/DataTable/DataTable.js]: @janhartmann - I forgot about that and totally agree. I think we need to re-think the design the implementation of useEffect hooks in DataTable. I also think we need to move these hooks into their own files/logic.

**Comment by jbetancur** [comment]
https://github.com/jbetancur/react-data-table-component/issues/335

Empty right now, but a placeholder for refactor work

File: src/DataTable/DataTable.js (hunk 3 above)
Comment (sarathdr) [reply src/DataTable/DataTable.js]: I thought about it as well, @jbetancur need to clean up that

**Comment by jbetancur** [comment]
Strange thing is the hooks linter does not catch it

**Comment by jbetancur** [comment]
I just mocked up an actual delete story and it seems to be working. @sarathdr  your logic solves the issue. Just need to get that test working. I'll have a look 

```

  useDidUpdateEffect(() => {
    if (pagination && paginationServer && data.length === 0 && currentPage > 1) {
      handleChangePage(currentPage - 1);
    }
  }, [paginationTotalRows]);
```

**Comment by jbetancur** [comment]
Actually, if you want to just remove the tests/snapshot updates I can merge this in. I can add the test in the refactor I am doing

**Comment by janhartmann** [comment]
Remember to put the “pagination” in the dependency array. Good rule is that all used variables within the hook needs to go into the dependency array, so if any of these change it will re-compute.

**Comment by janhartmann** [comment]
Or not sure if needed in this case

**Comment by sarathdr** [comment]
> Actually, if you want to just remove the tests/snapshot updates I can merge this in. I can add the test in the refactor I am doing

OK I will push after that

**Comment by jbetancur** [comment]
I created a codesandbox of what I am testing locally. It will server as a validation once 4.0.1 is cut

https://codesandbox.io/s/react-data-table-server-side-pagination-delete-xmdju

1. Nav to last page
2. delete both records
3. should nav to page 1

**Comment by sarathdr** [comment]
> I created a codesandbox of what I am testing locally. It will server as a validation once 4.0.1 is cut
> 
> https://codesandbox.io/s/react-data-table-server-side-pagination-delete-xmdju
> 
> 1. Nav to last page
> 2. delete both records
> 3. should nav to page 1

**Comment by jbetancur** [comment]
@janhartmann not needed in this case. While it is good practice in general, this sort of update is specific to paginationTotalRows changing (ie, a remote delete where the API returns some different row count. eg we don't want the page check to fire if say pagination is togged to false

**Comment by jbetancur** [comment]
@sarathdr Make sure you rebase form master as I just made some ordering changes to useEffect

**Comment by sarathdr** [comment]
> @sarathdr Make sure you rebase from master as I just made some ordering changes to useEffect

The pull request is closed, has this been merged ?

**Pull request closed without merging**